File: py/tests/world.py
"""Tests for the world.h bindings (no plugins required)."""
from __future__ import annotations

import mim

def test(world):
    b  = world.type_bool()
    i8 = world.type_i8()
    i8_7 = world.lit_i8(7)
    tt = world.lit_bool(True)
    ff = world.lit_bool(False)
    assert isinstance(i8_7, mim.Def)
    assert isinstance(world.lit_nat_0(), mim.Lit)
    assert isinstance(tt, mim.Lit)
    assert isinstance(ff, mim.Lit)
    assert isinstance(world.lit_ff(), mim.Lit)
    assert isinstance(world.lit_tt(), mim.Lit)
    assert isinstance(world.top_nat(), mim.Def)
    assert isinstance(b, mim.App)
    assert isinstance(i8, mim.App)
    assert isinstance(world.type_i32(), mim.App)
    assert isinstance(world.type_idx(world.top_nat()), mim.App)
    assert isinstance(world.type_idx(256), mim.App)
    idx_t = world.type_idx(world.top_nat())
    assert isinstance(world.lit(idx_t, 0), mim.Lit)
    assert isinstance(world.cn([b, i8]), mim.Pi)
    assert isinstance(world.arr(world.top_nat(), i8), mim.Arr)
    assert isinstance(world.mut_con([b, i8]), mim.Lam)


def test_mut_sigma_type_zero_overload(world):
    sigma = world.mut_sigma(2)

    sigma.set(0, world.type_nat())
    sigma.set(1, world.type_i32())
    assert sigma.num_ops() == 2


def test_implicit_app_single_argument_overload(world):
    from mim._plugins.option import option

    world.driver().load_plugins(["option"])
    result = world.implicit_app(
        world.annex(option.some.value), world.lit_nat(7)
    )

    assert isinstance(result, mim.Def)


def test_lam_set_body_with_filter_overload(world):
    identity = world.mut_lam(world.type_nat(), world.type_nat())

    identity.set_body(True, identity.var())

    assert identity.body() == identity.var()


def test_lit_i64_accepts_negative_values(world):
    value = world.lit_i64(-1)

    assert value.get_nat() == 2**64 - 1


def test_sym(world):
    s = world.sym("hello")
    assert isinstance(s, mim.Sym)
    assert s.str() == "hello"
    assert not s.empty()
    assert s.size() == len("hello")
    s = world.sym("")
    assert s.empty()
    assert s.size() == 0

File: py/bindings/world.cpp
#include <nanobind/nanobind.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/string_view.h>
#include <nanobind/stl/vector.h>

#include <cstdint>

#include <fe/sym.h>

#include <mim/def.h>
#include <mim/lam.h>
#include <mim/world.h>

#include <mim/pass/optimize.h>

namespace nb = nanobind;

namespace mim {

using DefVector = std::vector<const Def*>;

void init_world(nb::module_& m) {
    // clang-format off
    nb::class_<World>(m, "World", nb::never_destruct())
        .def("write", nb::overload_cast<>(&World::write))
        .def("top_nat", &World::top_nat, nb::rv_policy::reference_internal)
        .def("type_bool", &World::type_bool, nb::rv_policy::reference_internal)
        .def("type_i8", &World::type_i8, nb::rv_policy::reference_internal)
        .def("lit_nat_0", &World::lit_nat_0, nb::rv_policy::reference_internal)
        .def("lit_bool", &World::lit_bool, nb::rv_policy::reference_internal)
        .def("lit_ff", &World::lit_ff, nb::rv_policy::reference_internal)
        .def("lit_tt", &World::lit_tt, nb::rv_policy::reference_internal)
        .def("lit", &World::lit, nb::rv_policy::reference_internal)
        .def("type_idx", static_cast<const Def* (World::*)(const Def*)>(&World::type_idx), nb::rv_policy::reference_internal)
        .def("type_idx", static_cast<const Def* (World::*)(nat_t)>(&World::type_idx), nb::rv_policy::reference_internal)
        .def("cn", [](World& w, DefVector doms) { return w.cn(Defs(doms)); }, nb::rv_policy::reference_internal)
        .def("lit_i8", static_cast<const Lit* (World::*)(u8)>(&World::lit_i8), nb::rv_policy::reference_internal)
        .def("type_i32", &World::type_i32, nb::rv_policy::reference_internal)
        .def("sym", static_cast<fe::Sym (World::*)(std::string_view)>(&World::sym))
        .def("pi",      [](World& w, const Def* dom , const Def* codom ) { return w.pi     (      dom,       codom  ); }, nb::rv_policy::reference_internal)
        .def("pi",      [](World& w, const Def* dom , DefVector  codoms) { return w.pi     (      dom,  Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("pi",      [](World& w, DefVector  doms, const Def* codom ) { return w.pi     (Defs(doms),      codom  ); }, nb::rv_policy::reference_internal)
        .def("pi",      [](World& w, DefVector  doms, DefVector  codoms) { return w.pi     (Defs(doms), Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("mut_lam", [](World& w, const Def* dom , const Def* codom ) { return w.mut_lam(      dom,       codom  ); }, nb::rv_policy::reference_internal)
        .def("mut_lam", [](World& w, const Def* dom , DefVector  codoms) { return w.mut_lam(      dom,  Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("mut_lam", [](World& w, DefVector  doms, const Def* codom ) { return w.mut_lam(Defs(doms),      codom  ); }, nb::rv_policy::reference_internal)
        .def("mut_lam", [](World& w, DefVector  doms, DefVector  codoms) { return w.mut_lam(Defs(doms), Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("mut_fun", [](World& w, const Def* dom , const Def* codom ) { return w.mut_fun(      dom,       codom  ); }, nb::rv_policy::reference_internal)
        .def("mut_fun", [](World& w, const Def* dom , DefVector  codoms) { return w.mut_fun(      dom,  Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("mut_fun", [](World& w, DefVector  doms, const Def* codom ) { return w.mut_fun(Defs(doms),      codom  ); }, nb::rv_policy::reference_internal)
        .def("mut_fun", [](World& w, DefVector  doms, DefVector  codoms) { return w.mut_fun(Defs(doms), Defs(codoms)); }, nb::rv_policy::reference_internal)
        .def("mut_con", [](World& w, const Def* dom                    ) { return w.mut_con(     dom                ); }, nb::rv_policy::reference_internal)
        .def("mut_con", [](World& w, DefVector  doms                   ) { return w.mut_con(Defs(doms)              ); }, nb::rv_policy::reference_internal)
        .def(         "app", [](World& w, const Def* callee, const Def* arg ) { return w.         app(callee,      arg ) ; }, nb::rv_policy::reference_internal)
        .def(         "app", [](World& w, const Def* callee, DefVector  args) { return w.         app(callee, Defs(args)); }, nb::rv_policy::reference_internal)
        .def("implicit_app", [](World& w, const Def* callee, const Def* arg ) { return w.implicit_app(callee,      arg ) ; }, nb::rv_policy::reference_internal)
        .def("implicit_app", [](World& w, const Def* callee, DefVector  args) { return w.implicit_app(callee, Defs(args)); }, nb::rv_policy::reference_internal)
        .def("arr",     [](World& w, Def* arity, Def* body) { return w.arr(arity, body); }, nb::rv_policy::reference_internal)
        .def("optimize", [](World& w) { optimize(w); })
        .def("set", [](World& w, std::string name) { w.set(name); })
        .def("dot", static_cast<void (World::*)(const char*, bool, bool) const>(&World::dot))
        .def("annex", [](World& w, uint64_t id) { return w.annex(id); }, nb::rv_policy::reference_internal);
    // clang-format on
}

} // namespace mim
